### PR TITLE
Improved decision trees

### PR DESCRIPTION
**Pull request by Gaspi**
When computing the WHNF of a term to inspect its head while running a decision tree, the computed WHNF can be later discarded, either because the RHS of the matching rule doesn't refer to the computed subterms or because no rule fires at the head.

### CODE REVIEW

File: kernel/basic.mli [hunk 1]
```diff
@@ -117,6 +117,10 @@ val split : int -> 'a list -> 'a list * 'a list
 
 val rev_mapi : (int -> 'a -> 'b) -> 'a list -> 'b list
 
+val replace : ('a -> 'a) -> int -> 'a list -> 'a * 'a list
+(** [replace f l n] returns the list [l] where the [n]-th element, [e],
+    is replaced with [f e] together with [f e] itself. *)
+
```
Comment (GuillaumeGen) [inline kernel/basic.mli]: Comments are above the functions in this file.

File: kernel/basic.ml [hunk 2]
```diff
@@ -148,6 +148,14 @@ let rev_mapi f l =
...
   rmap_f 0 [] l
 
+let replace f =
+  let rec aux acc i = function
+    | []   -> failwith "replace"
+    | h::t ->
+      if i = 0
+      then let fh = f h in (fh, List.rev_append acc (fh::t))
+      else aux (h::acc) (i-1) t in
+  aux []
```
Comment (GuillaumeGen) [inline kernel/basic.ml]: This seems to be a function way too specific to be in Basic.

File: kernel/basic.ml [hunk 3]
```diff
@@ -148,6 +148,7 @@ let rev_mapi f l =
   in
   rmap_f 0 [] l
 
+let concat l1 = function [] -> l1 | l2 -> l1@l2
```
Comment (francoisthire) [inline kernel/basic.ml]: Maybe you should add a hint to say that this function should be inlined (with [@@ inline]

Comment (GuillaumeGen) [reply kernel/basic.ml]: List.append exists in the standard library. Is it a good idea to create (almost) an alias?
Is it really more efficient to add a test?

**Comment by francoisthire** [comment]
Either we bench this, otherwise I think we should keep Gaspard's definition.

Le mer. 17 juil. 2019 à 11:58, GuillaumeGen <notifications@github.com> a
écrit :

**Pull request merged**